### PR TITLE
ENH: Improve performance of DE's `_ensure_constraint`

### PR DESCRIPTION
**Pull request by mscavnicky**
Having profiled my script that uses `optimize.differential_evolution`, I was surprised that significant amount of time was spent in DE's internals rather than within my slow fitness function. The main culprit seems to be the `_ensure_constraint` method. This PR rewrites it with `np.where`.

I have compared the performance of the two methods using 3 array sizes (10, 100, 1000) and 3 ratios of out-of-bounds elements (0.1%, 0.5%, 1.0%). The results are as follows:

```
dim	ratio	method				result
10	0.001	ensure_constraint_scipy:	10000 loops, best of 3: 5.38 usec per loop
10	0.005	ensure_constraint_scipy:	10000 loops, best of 3: 5.23 usec per loop
10	0.01	ensure_constraint_scipy:	10000 loops, best of 3: 5.54 usec per loop
100	0.001	ensure_constraint_scipy:	10000 loops, best of 3: 44.1 usec per loop
100	0.005	ensure_constraint_scipy:	10000 loops, best of 3: 44.3 usec per loop
100	0.01	ensure_constraint_scipy:	10000 loops, best of 3: 44.4 usec per loop
1000	0.001	ensure_constraint_scipy:	10000 loops, best of 3: 444 usec per loop
1000	0.005	ensure_constraint_scipy:	10000 loops, best of 3: 454 usec per loop
1000	0.01	ensure_constraint_scipy:	10000 loops, best of 3: 452 usec per loop
10	0.001	ensure_constraint_where:	10000 loops, best of 3: 4.93 usec per loop
10	0.005	ensure_constraint_where:	10000 loops, best of 3: 4.88 usec per loop
10	0.01	ensure_constraint_where:	10000 loops, best of 3: 5.26 usec per loop
100	0.001	ensure_constraint_where:	10000 loops, best of 3: 4.67 usec per loop
100	0.005	ensure_constraint_where:	10000 loops, best of 3: 5.08 usec per loop
100	0.01	ensure_constraint_where:	10000 loops, best of 3: 5.05 usec per loop
1000	0.001	ensure_constraint_where:	10000 loops, best of 3: 5.7 usec per loop
1000	0.005	ensure_constraint_where:	10000 loops, best of 3: 5.52 usec per loop
1000	0.01	ensure_constraint_where:	10000 loops, best of 3: 6 usec per loop
```

Benchmarks were run on 
```
> python3 --version
Python 3.6.3
> sysctl -n machdep.cpu.brand_string
Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz
```

### CODE REVIEW

**Comment by andyfaff** [comment]
@mscavnicky thank you for contributing to scipy for the first time.
The test fail on travis is unrelated to this PR. 

Most of the problems I use `differential_evolution` for probably only have on the order of 10 parameters. At that level the difference between the two approaches you mention is negligible.
(my objective function normally takes an order of magnitude longer than any overhead incurred in scipy). Still, it's good to improve performance.

Unfortunately the improvement you suggest has a few issues with it. The main one is that for all the entries that are outside the range, it uses the same random number as the replacement.

```
import numpy as np
def f(trial):
    res = np.copy(trial)
    for index, param in enumerate(trial):
        if param > 1 or param < 0:
            res[index] = np.random.rand()
    return res

def f2(trial):
    res = np.copy(trial)
    for index in np.where((trial < 0) | (trial > 1)):
        res[index] = np.random.rand()
    return res

ll = -1
ul = 2
trial = (ul - ll) * np.random.rand((10)) + ll
print('trial', trial)
print('original', f(trial))
print('suggested', f2(trial)
```
gives the following:
```
trial [ 1.54373367  0.59144653 -0.12993146 -0.99878989 -0.61958958  0.47539014
  0.81057796  0.15478569 -0.02327803  1.16833128]
original [ 0.09154946  0.59144653  0.57734997  0.08478023  0.19129296  0.47539014
  0.81057796  0.15478569  0.5429328   0.70491116]
suggested [ 0.01449935  0.59144653  0.01449935  0.01449935  0.01449935  0.47539014
  0.81057796  0.15478569  0.01449935  0.01449935]
```
You can see in the last one that the replacement entries are all the same.

A better replacement for the code is:
```
def f3(trial):
      return np.where((trial < 0) | (trial > 1), np.random.rand(trial.size), trial)
```

which makes the scipy replacement:
```
trial = np.where((trial < 0) | (trial > 1), np.random.rand(trial.size), trial)
```

Thus, we can get rid of the for loop and the if test. I wonder how this changes the timings?

File: scipy/optimize/_differentialevolution.py [hunk 1]
```diff
@@ -689,9 +689,8 @@ def _ensure_constraint(self, trial):
         """
         make sure the parameters lie between the limits
         """
-        for index, param in enumerate(trial):
-            if param > 1 or param < 0:
-                trial[index] = self.random_number_generator.rand()
+        for index in np.where((trial < 0) | (trial > 1)):
```
Comment (andyfaff) [inline scipy/optimize/_differentialevolution.py]: The improvement you suggest has a few issues with it. The main one is that for all the entries that are outside the range, it uses the same random number as the replacement.
Use
```
trial = np.where((trial < 0) | (trial > 1), np.random.rand(trial.size), trial)
```
instead.

**Review by andyfaff — CHANGES_REQUESTED** [review]

**Comment by mscavnicky** [comment]
Thanks for catching the error. That was a basic mistake on my part.

The method that you have proposed seems to be slightly slower especially in the larger dimensions since it needs to allocate and generate the whole random array of the size of the trial.

```trial = np.where((trial < 0) | (trial > 1), np.random.rand(trial.size), trial)```

By replacing `np.where` with `np.putmask` we can get rid of the assignment at least. That is a micro-optimisation but seems to consistently perform better.

```np.putmask(trial, (trial < 0) | (trial > 1), np.random.rand(trial.size))```

The new timings are in the table below. The structure of the table is changed to make it more readable (every method now has its own column). I have also added extra rows for arrays with 2 elements.

```
dim	prob	scipy	np.where np.putmask
2	0.001	1.89	5.75	 4.51
2	0.005	2.15	5.98	 4.59
2	0.01	1.94	6.02	 5.03
10	0.001	5.23	5.83	 4.9
10	0.005	5.15	6.11	 4.82
10	0.01	5.22	5.89	 4.98
100	0.001	42.4	7.3	 6.13
100	0.005	42.5	7.03	 6.08
100	0.01	44.8	7.26	 6.04
1000	0.001	427	20	 19.3
1000	0.005	428	20.2	 18.8
1000	0.01	427	19.8	 18.7
```

The current implementation performs better in low dimensions which you mention are your general use case, but its complexity grows linearly. The method using `np.putmask` has some fixed costs but has a steady performance with growing dimension. My current problem has 50-100 dimensions and the ~5x speedup is helpful, given that this method is called for every instance.

Ultimately, this is a trade-off between keeping DE optimized for smaller dimensions or making it more robust towards n-dimensional problems and introducing some fixed cost.

**Comment by mscavnicky** [comment]
Actually, my original reimplementation was incorrectly using the result of `np.where` which returns a tuple. It can be fixed and gives better performance than the methods from my previous comment. See below:

```
def fixed(trial):
  for index in np.where((trial < 0) | (trial > 1))[0]:
    trial[index] = self.random_number_generator.rand()
```

```
dim	prob	scipy	np.where	np.putmask	fixed
2	0.001	1.96	6		4.66		4.39
2	0.005	1.93	6.07		4.76		4.36
2	0.01	1.98	5.84		4.69		4.39
10	0.001	5.29	6.25		4.83		4.39
10	0.005	5.18	6.46		4.71		4.57
10	0.01	5.2	6.36		4.82		4.58
100	0.001	42.8	7.23		5.97		4.44
100	0.005	45.8	7.23		6.03		4.46
100	0.01	42.9	7.31		5.86		4.41
1000	0.001	431	20.2		19.4		5.37
1000	0.005	430	20.5		18.6		5.51
1000	0.01	428	20.1		18.8		5.59
```

The conclusion of the previous comment still holds.

**Comment by andyfaff** [comment]
So it seems like the last suggestion is the way to go (although I like the one-liner). Could you change the PR to that please?

**Comment by andyfaff** [comment]
Optimisation applied in #8177

**Pull request closed without merging**